overlord/snapstate/snapstate_test.go

```diff
@@ -10969,246 +10967,6 @@ func (snapStateSuite) TestDefaultContentPlugProviders(c *C) {
 	c.Check(providers, DeepEquals, []string{"common-themes", "some-snap"})
 }
 
-type snapSetupSuite struct{}
-
-var _ = Suite(&snapSetupSuite{})
-
-type canRemoveSuite struct {
-	st        *state.State
-	deviceCtx snapstate.DeviceContext
-
-	bootloader *bootloadertest.MockBootloader
-}
-
-var _ = Suite(&canRemoveSuite{})
-
-func (s *canRemoveSuite) SetUpTest(c *C) {
-	dirs.SetRootDir(c.MkDir())
-	s.st = state.New(nil)
-	s.deviceCtx = &snapstatetest.TrivialDeviceContext{DeviceModel: DefaultModel()}
-
-	s.bootloader = bootloadertest.Mock("mock", c.MkDir())
-	bootloader.Force(s.bootloader)
-}
-
-func (s *canRemoveSuite) TearDownTest(c *C) {
-	dirs.SetRootDir("/")
-	bootloader.Force(nil)
-}
-
-func (s *canRemoveSuite) TestAppAreAlwaysOKToRemove(c *C) {
-	info := &snap.Info{
-		SnapType: snap.TypeApp,
-	}
-	info.RealName = "foo"
-
-	c.Check(snapstate.CanRemove(s.st, info, &snapstate.SnapState{Active: true}, false, s.deviceCtx), Equals, true)
-	c.Check(snapstate.CanRemove(s.st, info, &snapstate.SnapState{Active: true}, true, s.deviceCtx), Equals, true)
-}
-
-func (s *canRemoveSuite) TestLastGadgetsAreNotOK(c *C) {
-	info := &snap.Info{
-		SnapType: snap.TypeGadget,
-	}
-	info.RealName = "foo"
-
-	c.Check(snapstate.CanRemove(s.st, info, &snapstate.SnapState{}, true, s.deviceCtx), Equals, false)
-}
-
-func (s *canRemoveSuite) TestLastOSAndKernelAreNotOK(c *C) {
-	os := &snap.Info{
-		SnapType: snap.TypeOS,
-	}
-	os.RealName = "os"
-	kernel := &snap.Info{
-		SnapType: snap.TypeKernel,
-	}
-	// this kernel part of the model
-	kernel.RealName = "kernel"
-
-	c.Check(snapstate.CanRemove(s.st, os, &snapstate.SnapState{}, true, s.deviceCtx), Equals, false)
-
-	c.Check(snapstate.CanRemove(s.st, kernel, &snapstate.SnapState{}, true, s.deviceCtx), Equals, false)
-}
-
-func (s *canRemoveSuite) TestKernelBootInUseIsKept(c *C) {
-	kernel := &snap.Info{
-		SnapType: snap.TypeKernel,
-		SideInfo: snap.SideInfo{
-			Revision: snap.R(3),
-		},
-	}
-	kernel.RealName = "kernel"
-
-	s.bootloader.SetBootKernel(fmt.Sprintf("%s_%s.snap", kernel.RealName, kernel.SideInfo.Revision))
-
-	removeAll := false
-	c.Check(snapstate.CanRemove(s.st, kernel, &snapstate.SnapState{}, removeAll, s.deviceCtx), Equals, false)
-}
-
-func (s *canRemoveSuite) TestOstInUseIsKept(c *C) {
-	base := &snap.Info{
-		SnapType: snap.TypeBase,
-		SideInfo: snap.SideInfo{
-			Revision: snap.R(3),
-		},
-	}
-	base.RealName = "core18"
-
-	s.bootloader.SetBootBase(fmt.Sprintf("%s_%s.snap", base.RealName, base.SideInfo.Revision))
-
-	removeAll := false
-	c.Check(snapstate.CanRemove(s.st, base, &snapstate.SnapState{}, removeAll, s.deviceCtx), Equals, false)
-}
-
-func (s *canRemoveSuite) TestRemoveNonModelKernelIsOk(c *C) {
-	kernel := &snap.Info{
-		SnapType: snap.TypeKernel,
-	}
-	kernel.RealName = "other-non-model-kernel"
-
-	c.Check(snapstate.CanRemove(s.st, kernel, &snapstate.SnapState{}, true, s.deviceCtx), Equals, true)
-}
-
-func (s *canRemoveSuite) TestRemoveNonModelKernelStillInUseNotOk(c *C) {
```

Comment:
we don't seem to have a test like this anymore

Reply:
because you convinced me that the test was testing an impossible situation?

this is the "the kernel used for boot is not the kernel in the model" scenario

Reply:
I see you are right, but then the current code independently of whether that's plausible or not as a strange property:

it won't let you remove the single revision of a kernel not in the model
but it will let you remove the whole thing
that is confusing on face value. Maybe we do need the more complicated code for consistency reasons, and belt and suspender reasons.